lib/benchmark/bench.go

```diff
@@ -52,98 +63,93 @@ type BenchmarkResult struct {
-	go func() {
-		interval := time.Duration(float64(1) / float64(bench.Rate) * float64(time.Second))
-		ticker := time.NewTicker(interval)
-		defer ticker.Stop()
-		for {
-			select {
-			case <-ticker.C:
-				// notice how we start the timer regardless of whether any goroutine can process it
-				// this is to account for coordinated omission,
-				// http://psy-lob-saw.blogspot.com/2015/03/fixing-ycsb-coordinated-omission.html
-				measure := &benchMeasure{
-					Start: time.Now(),
-				}
-				select {
-				case requestC <- measure:
-				case <-ctx.Done():
-					return
-				}
-			case <-ctx.Done():
-				return
-			}
-		}
-	}()
+	go produceMeasures(ctx, c.Rate, resultC)
 
-	var result BenchmarkResult
-	// from one millisecond to 60000 milliseconds (minute) with 3 digits precision
-	result.Histogram = hdrhistogram.New(1, 60000, 3)
+	var result Result
+	// from one millisecond to 60000 milliseconds (minute) with 3 digits precision, refer to constants
+	result.Histogram = hdrhistogram.New(MinValue, MaxValue, SignificantFigures)
+	results := make([]*benchMeasure, 0, c.MinimumMeasurements)
+	statusTicker := time.NewTicker(1 * time.Second)
+	timeElapsed := false
+	start := time.Now()
 
-	var doneThreads int
-	var timeoutC <-chan time.Time
-	doneC := ctx.Done()
 	for {
+		if c.MinimumWindow <= time.Since(start) {
+			timeElapsed = true
+		}
 		select {
-		case <-timeoutC:
-			result.LastError = trace.BadParameter("several requests hang: timeout waiting for %v threads to finish", bench.Threads-doneThreads)
-			return &result, nil
-		case <-doneC:
-			// give it a couple of seconds to wrap up the goroutines,
-			// set up the timer that will fire up if the all goroutines were not finished
-			doneC = nil
-			waitTime := time.Duration(result.Histogram.Max()) * time.Millisecond
-			// going to wait latency + buffer to give requests in flight to wrap up
-			waitTime = time.Duration(1.2 * float64(waitTime))
-			timeoutC = time.After(waitTime)
 		case measure := <-responseC:
-			if measure.ThreadCompleted {
-				doneThreads++
-				if doneThreads == bench.Threads {
-					return &result, nil
-				}
-			} else {
-				if measure.Error != nil {
-					result.RequestsFailed++
-					result.LastError = measure.Error
-				}
-				result.RequestsOriginated++
-				result.Histogram.RecordValue(int64(measure.End.Sub(measure.Start) / time.Millisecond))
+			result.Histogram.RecordValue(int64(measure.End.Sub(measure.Start) / time.Millisecond))
+			results = append(results, measure)
+			if timeElapsed && len(results) >= c.MinimumMeasurements {
+				go cancelWorkers()
+			}
+			if measure.Error != nil {
+				result.RequestsFailed++
+				result.LastError = measure.Error
 			}
+			result.RequestsOriginated++
+		case <-workerCtx.Done():
+			result.Duration = time.Since(start)
+			return &result, nil
+		case <-statusTicker.C:
+			log.Printf("working... observations: %d", len(results))
 		}
 	}
+}
+
+
+
+func produceMeasures(ctx context.Context, rate int, c chan<- *benchMeasure) {
```

Comment:
I would use `measuresCh chan benchMeasure` instead of pointer here, in Go it could be faster to use values allocated on stack:

https://segment.com/blog/allocation-efficiency-in-high-performance-go-services/